TestsFoundation/TestsFoundation/Extensions/XCUIElementExtensions.swift

```diff
@@ -19,50 +19,241 @@
+        tap()
+        return self
+    }
+
+    @discardableResult
+    func waitUntil(condition: ElementCondition,
+                   expected: String? = nil,
+                   timeout: TimeInterval = defaultTimeout,
+                   gracePeriod: UInt32 = defaultGracePeriod) -> XCUIElement {
+        tacticalSleep()
+        let deadline = Date().addingTimeInterval(timeout)
+        while Date() < deadline {
+            var result = false
+
+            switch condition {
+            case .vanish:
+                result = isVanished
+            case .visible:
+                result = isVisible
+            case .value:
+                result = hasValue(value: expected!)
+            case .label:
+                result = hasLabel(label: expected!)
+            case .enabled:
+                result = exists && isEnabled
+            case .selected:
+                result = exists && isSelected
+            case .unselected:
+                result = !isSelected
+            case .hittable:
+                result = isHittable
+            case .labelContaining:
+                result = label.contains(expected!)
+            case .labelHasPrefix:
+                result = label.hasPrefix(expected!)
+            }
+            if result { break } else { sleep(gracePeriod) }
+        }
+        return self
+    }
+
+    @discardableResult
+    func actionUntilElementCondition(action: ElementAction = .tap,
+                                     element: XCUIElement? = nil,
+                                     condition: ElementCondition,
+                                     expected: String? = nil,
+                                     timeout: TimeInterval = defaultTimeout,
+                                     gracePeriod: UInt32 = defaultGracePeriod) -> Bool {
+        tacticalSleep()
+        let deadline = Date().addingTimeInterval(timeout)
+        while Date() < deadline {
+            let actualElement = element ?? self
+            var result = false
+
+            switch condition {
+            case .vanish:
+                result = actualElement.isVanished
+            case .visible:
+                result = actualElement.isVisible
+            case .value:
+                result = actualElement.hasValue(value: expected!)
+            case .label:
+                result = actualElement.hasLabel(label: expected!)
+            case .enabled:
+                result = actualElement.exists && actualElement.isEnabled
+            case .selected:
+                result = actualElement.exists && actualElement.isSelected
+            case .unselected:
+                result = !actualElement.isSelected
+            case .hittable:
+                result = actualElement.isHittable
+            case .labelContaining:
+                result = label.contains(expected!)
+            case .labelHasPrefix:
+                result = label.hasPrefix(expected!)
+            }
+            if result { return true } else {
```

Comment:
Use `guard` for the early exit and the switch can be indented one level to the left.

Reply:
Done.